### ISSUE TITLE
Infinite loop 'while( each { 1=>'a' } )'

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14706]
Migrated from [rt.perl.org#125205](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=125205) (status was 'rejected')

Searchable as RT125205$

**Comment by p5pRT** [comment]
### From @KES777

This is a bug report for perl from kes\-kes@&#8203;yandex\.ru\,
generated with the help of perlbug 1\.39 running under perl 5\.18\.2\.

\#\!/usr/bin/perl

use Modern&#8203;::Perl;

my $val = \{ 1 => 'a' \};

\#test 1 OK
while\( my\( $key\, $value \) =  each %\{ $val \} \) \{
&nbsp;    say "$key $value";
\}

\#test 2 OK \(Why this work if "Type of argument to each on reference must be unblessed hashref or arrayref at \./test\.pl line 18\."
while\( my\( $key\, $value \) =  each %$val \) \{
&nbsp;    say "$key $value";
\}

\#test 3 FAIL not Expected
\#while\( my\( $key\, $value \) =  each \( 1 => 'a' \) \) \{
\#    say "$key $value";
\#\}

\#test4 OK as Expected \(see perldoc each\)
my %hash =  %\{ \{ 1 => 'a' \} \}; \#same hash as in prev example&#8203;: %hash =  \( 1 => 'a' \)
while\( my\( $key\, $value \) =  each %hash \) \{
&nbsp;    say "$key $value";
\}

\#test 5 FAIL not expected \(infinite loop\)
while\( my\( $key\, $value \) =  each %\{ \{ 1 => 'a' \} \} \) \{
&nbsp;    say "$key $value";
\}

\#test 6 FAIL not Expected \(infinite loop\)
while\( my\( $key\, $value \) =  each \{ 1 => 'a' \} \) \{
&nbsp;    say "$key $value";
\}


\_\_END\_\_

May you tell me please where I can view status of this BUG report?



---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=critical


---
Site configuration information for perl 5\.18\.2&#8203;:

Configured by Debian Project at Thu Mar 27 18&#8203;:28&#8203;:21 UTC 2014\.

Summary of my perl5 \(revision 5 version 18 subversion 2\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.2\.0\-58\-generic\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux brownie 3\.2\.0\-58\-generic \#88\-ubuntu smp tue dec 3 17&#8203;:37&#8203;:58 utc 2013 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-D\_FORTIFY\_SOURCE=2 \-g \-O2 \-fstack\-protector \-\-param=ssp\-buffer\-size=4 \-Wformat \-Werror=format\-security \-Dldflags= \-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dlddlflags=\-shared \-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.18 \-Darchlib=/usr/lib/perl/5\.18 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.18\.2 \-Dsitearch=/usr/local/lib/perl/5\.18\.2 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-Ui\_libutil \-Uversiononly \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.18\.2 \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fstack\-protector \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fstack\-protector \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.8\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.18\.2
&nbsp;    gnulibc\_version='2\.19'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS default for modules installed from CPAN\.
&nbsp;    DEBPKG&#8203;:debian/db\_file\_ver \- http&#8203;://bugs\.debian\.org/340047 Remove overly restrictive DB\_File version check\.
&nbsp;    DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with Debian\-specific information\.
&nbsp;    DEBPKG&#8203;:debian/enc2xs\_inc \- http&#8203;://bugs\.debian\.org/290336 Tweak enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
&nbsp;    DEBPKG&#8203;:debian/errno\_ver \- http&#8203;://bugs\.debian\.org/343351 Remove Errno version check due to upgrade problems with long\-running processes\.
&nbsp;    DEBPKG&#8203;:debian/libperl\_embed\_doc \- http&#8203;://bugs\.debian\.org/186778 Note that libperl\-dev package is required for embedded linking
&nbsp;    DEBPKG&#8203;:fixes/respect\_umask \- Respect umask during installation
&nbsp;    DEBPKG&#8203;:debian/writable\_site\_dirs \- Set umask approproately for site install directories
&nbsp;    DEBPKG&#8203;:debian/extutils\_set\_libperl\_path \- EU&#8203;:MM&#8203;: Set location of libperl\.a to /usr/lib
&nbsp;    DEBPKG&#8203;:debian/no\_packlist\_perllocal \- Don't install \.packlist or perllocal\.pod for perl or vendor
&nbsp;    DEBPKG&#8203;:debian/prefix\_changes \- Fiddle with \*PREFIX and variables written to the makefile
&nbsp;    DEBPKG&#8203;:debian/fakeroot \- Postpone LD\_LIBRARY\_PATH evaluation to the binary targets\.
&nbsp;    DEBPKG&#8203;:debian/instmodsh\_doc \- Debian policy doesn't install \.packlist files for core or vendor\.
&nbsp;    DEBPKG&#8203;:debian/ld\_run\_path \- Remove standard libs from LD\_RUN\_PATH as per Debian policy\.
&nbsp;    DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to /etc/perl/Net as /usr may not be writable\.
&nbsp;    DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
&nbsp;    DEBPKG&#8203;:debian/module\_build\_man\_extensions \- http&#8203;://bugs\.debian\.org/479460 Adjust Module&#8203;::Build manual page extensions for the Debian Perl policy
&nbsp;    DEBPKG&#8203;:debian/prune\_libs \- http&#8203;://bugs\.debian\.org/128355 Prune the list of libraries wanted to what we actually need\.
&nbsp;    DEBPKG&#8203;:fixes/net\_smtp\_docs \- \[rt\.cpan\.org \#36038\] http&#8203;://bugs\.debian\.org/100195 Document the Net&#8203;::SMTP 'Port' option
&nbsp;    DEBPKG&#8203;:debian/perlivp \- http&#8203;://bugs\.debian\.org/510895 Make perlivp skip include directories in /usr/local
&nbsp;    DEBPKG&#8203;:debian/cpanplus\_definstalldirs \- http&#8203;://bugs\.debian\.org/533707 Configure CPANPLUS to use the site directories by default\.
&nbsp;    DEBPKG&#8203;:debian/cpanplus\_config\_path \- Save local versions of CPANPLUS&#8203;::Config&#8203;::System into /etc/perl\.
&nbsp;    DEBPKG&#8203;:debian/deprecate\-with\-apt \- http&#8203;://bugs\.debian\.org/702096 Point users to Debian packages of deprecated core modules
&nbsp;    DEBPKG&#8203;:debian/squelch\-locale\-warnings \- http&#8203;://bugs\.debian\.org/508764 Squelch locale warnings in Debian package maintainer scripts
&nbsp;    DEBPKG&#8203;:debian/skip\-upstream\-git\-tests \- Skip tests specific to the upstream Git repository
&nbsp;    DEBPKG&#8203;:debian/patchlevel \- http&#8203;://bugs\.debian\.org/567489 List packaged patches for 5\.18\.2\-2ubuntu1 in patchlevel\.h
&nbsp;    DEBPKG&#8203;:debian/skip\-kfreebsd\-crash \- http&#8203;://bugs\.debian\.org/628493 \[perl \#96272\] Skip a crashing test case in t/op/threads\.t on GNU/kFreeBSD
&nbsp;    DEBPKG&#8203;:fixes/document\_makemaker\_ccflags \- http&#8203;://bugs\.debian\.org/628522 \[rt\.cpan\.org \#68613\] Document that CCFLAGS should include $Config\{ccflags\}
&nbsp;    DEBPKG&#8203;:debian/find\_html2text \- http&#8203;://bugs\.debian\.org/640479 Configure CPAN&#8203;::Distribution with correct name of html2text
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_stack \- http&#8203;://bugs\.debian\.org/650175 Disable failing GNU/Hurd tests dist/threads/t/stack\.t
&nbsp;    DEBPKG&#8203;:fixes/manpage\_name\_Test\-Harness \- http&#8203;://bugs\.debian\.org/650451 \[rt\.cpan\.org \#73399\] cpan/Test\-Harness&#8203;: add NAME headings in modules with POD
&nbsp;    DEBPKG&#8203;:debian/makemaker\-pasthru \- http&#8203;://bugs\.debian\.org/660195 \[rt\.cpan\.org \#28632\] Make EU&#8203;::MM pass LD through to recursive Makefile\.PL invocations
&nbsp;    DEBPKG&#8203;:debian/perl5db\-x\-terminal\-emulator\.patch \- http&#8203;://bugs\.debian\.org/668490 Invoke x\-terminal\-emulator rather than xterm in perl5db\.pl
&nbsp;    DEBPKG&#8203;:debian/cpan\-missing\-site\-dirs \- http&#8203;://bugs\.debian\.org/688842 Fix CPAN&#8203;::FirstTime defaults with nonexisting site dirs if a parent is writable
&nbsp;    DEBPKG&#8203;:fixes/memoize\_storable\_nstore \- \[rt\.cpan\.org \#77790\] http&#8203;://bugs\.debian\.org/587650 Memoize&#8203;::Storable&#8203;: respect 'nstore' option not respected
&nbsp;    DEBPKG&#8203;:fixes/net\_ftp\_failed\_command \- \[rt\.cpan\.org \#37700\] http&#8203;://bugs\.debian\.org/491062 Net&#8203;::FTP&#8203;: cope gracefully with a failed command
&nbsp;    DEBPKG&#8203;:fixes/perlbug\-patchlist \- \[3541c11\] http&#8203;://bugs\.debian\.org/710842 \[perl \#118433\] Make perlbug look up the list of local patches at run time
&nbsp;    DEBPKG&#8203;:fixes/module\_metadata\_security\_doc \- \[68cdd4b\] CVE\-2013\-1437 documentation fix
&nbsp;    DEBPKG&#8203;:fixes/module\_metadata\_taint\_fix \- \[bff978f\] http&#8203;://bugs\.debian\.org/722210 \[rt\.cpan\.org \#88576\] untaint version\, if needed\, in Module&#8203;::Metadata
&nbsp;    DEBPKG&#8203;:fixes/IPC\-SysV\-spelling \- http&#8203;://bugs\.debian\.org/730558 \[rt\.cpan\.org \#86736\] Fix spelling of IPC\_CREAT in IPC\-SysV documentation
&nbsp;    DEBPKG&#8203;:fixes/fix\-undef\-source \-



---
@&#8203;INC for perl 5\.18\.2&#8203;:
&nbsp;    /home/kes/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    /home/kes/perl5/lib/perl5
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.18\.2
&nbsp;    /usr/local/share/perl/5\.18\.2
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.18
&nbsp;    /usr/share/perl/5\.18
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.18\.2&#8203;:
&nbsp;    HOME=/home/kes
&nbsp;    LANG=ru\_UA\.UTF\-8
&nbsp;    LANGUAGE=en
&nbsp;    LC\_MESSAGES=en\_US\.UTF\-8
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/kes/bin&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/sbin&#8203;:/bin&#8203;:/usr/games&#8203;:/usr/local/games
&nbsp;    PERL5LIB=/home/kes/perl5/lib/perl5
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_MB\_OPT=\-\-install\_base "/home/kes/perl5"
&nbsp;    PERL\_MM\_OPT=INSTALL\_BASE=/home/kes/perl5
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @KES777

Subject&#8203;: Infinite loop 'while\( each \{ 1=>'a' \} \)
Message\-Id&#8203;: \<5\.18\.2\_3607\_1431762613@&#8203;keswork>
Reply\-To&#8203;: kes\-kes@&#8203;yandex\.ru
To&#8203;: perlbug@&#8203;perl\.org
From&#8203;: kes\-kes@&#8203;yandex\.ru


This is a bug report for perl from kes\-kes@&#8203;yandex\.ru\,
generated with the help of perlbug 1\.39 running under perl 5\.18\.2\.

\#\!/usr/bin/perl

use Modern&#8203;::Perl;

my $val = \{ 1 => 'a' \};

\#test 1 OK
while\( my\( $key\, $value \) =  each %\{ $val \} \) \{
&nbsp;    say "$key $value";
\}

\#test 2 OK \(Why this work if "Type of argument to each on reference must be unblessed hashref or arrayref at \./test\.pl line 18\."
while\( my\( $key\, $value \) =  each %$val \) \{
&nbsp;    say "$key $value";
\}

\#test 3 FAIL not Expected
\#while\( my\( $key\, $value \) =  each \( 1 => 'a' \) \) \{
\#    say "$key $value";
\#\}

\#test4 OK as Expected \(see perldoc each\)
my %hash =  %\{ \{ 1 => 'a' \} \}; \#same hash as in prev example&#8203;: %hash =  \( 1 => 'a' \)
while\( my\( $key\, $value \) =  each %hash \) \{
&nbsp;    say "$key $value";
\}

\#test 5 FAIL not expected \(infinite loop\)
while\( my\( $key\, $value \) =  each %\{ \{ 1 => 'a' \} \} \) \{
&nbsp;    say "$key $value";
\}

\#test 6 FAIL not Expected \(infinite loop\)
while\( my\( $key\, $value \) =  each \{ 1 => 'a' \} \) \{
&nbsp;    say "$key $value";
\}


\_\_END\_\_

May you tell me please where I can view status of this BUG report?
\-\-\-
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=critical
\-\-\-
Site configuration information for perl 5\.18\.2&#8203;:

Configured by Debian Project at Thu Mar 27 18&#8203;:28&#8203;:21 UTC 2014\.

Summary of my perl5 \(revision 5 version 18 subversion 2\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.2\.0\-58\-generic\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux brownie 3\.2\.0\-58\-generic \#88\-ubuntu smp tue dec 3 17&#8203;:37&#8203;:58 utc 2013 x86\_64 x86\_64 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-D\_FORTIFY\_SOURCE=2 \-g \-O2 \-fstack\-protector \-\-param=ssp\-buffer\-size=4 \-Wformat \-Werror=format\-security \-Dldflags= \-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dlddlflags=\-shared \-Wl\,\-Bsymbolic\-functions \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.18 \-Darchlib=/usr/lib/perl/5\.18 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.18\.2 \-Dsitearch=/usr/local/lib/perl/5\.18\.2 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-Ui\_libutil \-Uversiononly \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.18\.2 \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fstack\-protector \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fstack\-protector \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.8\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.18\.2
&nbsp;    gnulibc\_version='2\.19'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS default for modules installed from CPAN\.
&nbsp;    DEBPKG&#8203;:debian/db\_file\_ver \- http&#8203;://bugs\.debian\.org/340047 Remove overly restrictive DB\_File version check\.
&nbsp;    DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with Debian\-specific information\.
&nbsp;    DEBPKG&#8203;:debian/enc2xs\_inc \- http&#8203;://bugs\.debian\.org/290336 Tweak enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
&nbsp;    DEBPKG&#8203;:debian/errno\_ver \- http&#8203;://bugs\.debian\.org/343351 Remove Errno version check due to upgrade problems with long\-running processes\.
&nbsp;    DEBPKG&#8203;:debian/libperl\_embed\_doc \- http&#8203;://bugs\.debian\.org/186778 Note that libperl\-dev package is required for embedded linking
&nbsp;    DEBPKG&#8203;:fixes/respect\_umask \- Respect umask during installation
&nbsp;    DEBPKG&#8203;:debian/writable\_site\_dirs \- Set umask approproately for site install directories
&nbsp;    DEBPKG&#8203;:debian/extutils\_set\_libperl\_path \- EU&#8203;:MM&#8203;: Set location of libperl\.a to /usr/lib
&nbsp;    DEBPKG&#8203;:debian/no\_packlist\_perllocal \- Don't install \.packlist or perllocal\.pod for perl or vendor
&nbsp;    DEBPKG&#8203;:debian/prefix\_changes \- Fiddle with \*PREFIX and variables written to the makefile
&nbsp;    DEBPKG&#8203;:debian/fakeroot \- Postpone LD\_LIBRARY\_PATH evaluation to the binary targets\.
&nbsp;    DEBPKG&#8203;:debian/instmodsh\_doc \- Debian policy doesn't install \.packlist files for core or vendor\.
&nbsp;    DEBPKG&#8203;:debian/ld\_run\_path \- Remove standard libs from LD\_RUN\_PATH as per Debian policy\.
&nbsp;    DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to /etc/perl/Net as /usr may not be writable\.
&nbsp;    DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
&nbsp;    DEBPKG&#8203;:debian/module\_build\_man\_extensions \- http&#8203;://bugs\.debian\.org/479460 Adjust Module&#8203;::Build manual page extensions for the Debian Perl policy
&nbsp;    DEBPKG&#8203;:debian/prune\_libs \- http&#8203;://bugs\.debian\.org/128355 Prune the list of libraries wanted to what we actually need\.
&nbsp;    DEBPKG&#8203;:fixes/net\_smtp\_docs \- \[rt\.cpan\.org \#36038\] http&#8203;://bugs\.debian\.org/100195 Document the Net&#8203;::SMTP 'Port' option
&nbsp;    DEBPKG&#8203;:debian/perlivp \- http&#8203;://bugs\.debian\.org/510895 Make perlivp skip include directories in /usr/local
&nbsp;    DEBPKG&#8203;:debian/cpanplus\_definstalldirs \- http&#8203;://bugs\.debian\.org/533707 Configure CPANPLUS to use the site directories by default\.
&nbsp;    DEBPKG&#8203;:debian/cpanplus\_config\_path \- Save local versions of CPANPLUS&#8203;::Config&#8203;::System into /etc/perl\.
&nbsp;    DEBPKG&#8203;:debian/deprecate\-with\-apt \- http&#8203;://bugs\.debian\.org/702096 Point users to Debian packages of deprecated core modules
&nbsp;    DEBPKG&#8203;:debian/squelch\-locale\-warnings \- http&#8203;://bugs\.debian\.org/508764 Squelch locale warnings in Debian package maintainer scripts
&nbsp;    DEBPKG&#8203;:debian/skip\-upstream\-git\-tests \- Skip tests specific to the upstream Git repository
&nbsp;    DEBPKG&#8203;:debian/patchlevel \- http&#8203;://bugs\.debian\.org/567489 List packaged patches for 5\.18\.2\-2ubuntu1 in patchlevel\.h
&nbsp;    DEBPKG&#8203;:debian/skip\-kfreebsd\-crash \- http&#8203;://bugs\.debian\.org/628493 \[perl \#96272\] Skip a crashing test case in t/op/threads\.t on GNU/kFreeBSD
&nbsp;    DEBPKG&#8203;:fixes/document\_makemaker\_ccflags \- http&#8203;://bugs\.debian\.org/628522 \[rt\.cpan\.org \#68613\] Document that CCFLAGS should include $Config\{ccflags\}
&nbsp;    DEBPKG&#8203;:debian/find\_html2text \- http&#8203;://bugs\.debian\.org/640479 Configure CPAN&#8203;::Distribution with correct name of html2text
&nbsp;    DEBPKG&#8203;:debian/hurd\_test\_skip\_stack \- http&#8203;://bugs\.debian\.org/650175 Disable failing GNU/Hurd tests dist/threads/t/stack\.t
&nbsp;    DEBPKG&#8203;:fixes/manpage\_name\_Test\-Harness \- http&#8203;://bugs\.debian\.org/650451 \[rt\.cpan\.org \#73399\] cpan/Test\-Harness&#8203;: add NAME headings in modules with POD
&nbsp;    DEBPKG&#8203;:debian/makemaker\-pasthru \- http&#8203;://bugs\.debian\.org/660195 \[rt\.cpan\.org \#28632\] Make EU&#8203;::MM pass LD through to recursive Makefile\.PL invocations
&nbsp;    DEBPKG&#8203;:debian/perl5db\-x\-terminal\-emulator\.patch \- http&#8203;://bugs\.debian\.org/668490 Invoke x\-terminal\-emulator rather than xterm in perl5db\.pl
&nbsp;    DEBPKG&#8203;:debian/cpan\-missing\-site\-dirs \- http&#8203;://bugs\.debian\.org/688842 Fix CPAN&#8203;::FirstTime defaults with nonexisting site dirs if a parent is writable
&nbsp;    DEBPKG&#8203;:fixes/memoize\_storable\_nstore \- \[rt\.cpan\.org \#77790\] http&#8203;://bugs\.debian\.org/587650 Memoize&#8203;::Storable&#8203;: respect 'nstore' option not respected
&nbsp;    DEBPKG&#8203;:fixes/net\_ftp\_failed\_command \- \[rt\.cpan\.org \#37700\] http&#8203;://bugs\.debian\.org/491062 Net&#8203;::FTP&#8203;: cope gracefully with a failed command
&nbsp;    DEBPKG&#8203;:fixes/perlbug\-patchlist \- \[3541c11\] http&#8203;://bugs\.debian\.org/710842 \[perl \#118433\] Make perlbug look up the list of local patches at run time
&nbsp;    DEBPKG&#8203;:fixes/module\_metadata\_security\_doc \- \[68cdd4b\] CVE\-2013\-1437 documentation fix
&nbsp;    DEBPKG&#8203;:fixes/module\_metadata\_taint\_fix \- \[bff978f\] http&#8203;://bugs\.debian\.org/722210 \[rt\.cpan\.org \#88576\] untaint version\, if needed\, in Module&#8203;::Metadata
&nbsp;    DEBPKG&#8203;:fixes/IPC\-SysV\-spelling \- http&#8203;://bugs\.debian\.org/730558 \[rt\.cpan\.org \#86736\] Fix spelling of IPC\_CREAT in IPC\-SysV documentation
&nbsp;    DEBPKG&#8203;:fixes/fix\-undef\-source \-

\-\-\-
@&#8203;INC for perl 5\.18\.2&#8203;:
&nbsp;    /home/kes/perl5/lib/perl5/x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    /home/kes/perl5/lib/perl5
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.18\.2
&nbsp;    /usr/local/share/perl/5\.18\.2
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.18
&nbsp;    /usr/share/perl/5\.18
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.

\-\-\-
Environment for perl 5\.18\.2&#8203;:
&nbsp;    HOME=/home/kes
&nbsp;    LANG=ru\_UA\.UTF\-8
&nbsp;    LANGUAGE=en
&nbsp;    LC\_MESSAGES=en\_US\.UTF\-8
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/kes/bin&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/sbin&#8203;:/bin&#8203;:/usr/games&#8203;:/usr/local/games
&nbsp;    PERL5LIB=/home/kes/perl5/lib/perl5
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_MB\_OPT=\-\-install\_base "/home/kes/perl5"
&nbsp;    PERL\_MM\_OPT=INSTALL\_BASE=/home/kes/perl5
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @KES777

[test.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1349400/720807/test.pl)

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat May 16 00&#8203;:56&#8203;:18 2015\, kes\-kes@&#8203;yandex\.ua wrote&#8203;:
> This is a bug report for perl from kes\-kes@&#8203;yandex\.ru\,
> generated with the help of perlbug 1\.39 running under perl 5\.18\.2\.
> 
> \#\!/usr/bin/perl
> 
> use Modern&#8203;::Perl;

To report a possible error in Perl itself\, you should submit the bug report in a manner in which other possible causes of the error are excluded\.  Modern&#8203;::Perl is a library which\, if I recall correctly\, tries to provide a lot of functionality which has become popular among Perl programmers in the last decade\, but it is nonetheless a library which might handle situations differently from Perl itself\.

So\, for the balance of my response\, I'm going to proceed as if you had \*not\* said 'use Modern&#8203;::Perl'\.  But I will assume at least Perl 5\.18\.2 \-\- the version from which you are filing this bug report \-\- and I will assume 'use strict;' and 'use warnings;'\.

> 
> my $val = \{ 1 => 'a' \};
> 
> \#test 1 OK
> while\( my\( $key\, $value \) =  each %\{ $val \} \) \{
>     say "$key $value";
> \}
> 
> \#test 2 OK \(Why this work if "Type of argument to each on reference
> must be unblessed hashref or arrayref at \./test\.pl line 18\."
> while\( my\( $key\, $value \) =  each %$val \) \{
>     say "$key $value";
> \}
> 
> \#test 3 FAIL not Expected
> \#while\( my\( $key\, $value \) =  each \( 1 => 'a' \) \) \{
> \#    say "$key $value";
> \#\}

This first gets a warning and then a failure message&#8203;:
\#\#\#\#\#
each on reference is experimental at 125205\.pl line 18\.

Type of argument to each on reference must be unblessed hashref or arrayref at 125205\.pl line 18\.
\#\#\#\#\#

See&#8203;: 'perldoc \-f each'\.  The way you are using 'each' above is neither on a hash nor on an array but on its third\, experimental version&#8203;:  "a scalar EXPR\, which must hold reference to an unblessed hash or array\."  Your expression does not hold a reference; hence the error\.


> 
> \#test4 OK as Expected \(see perldoc each\)
> my %hash =  %\{ \{ 1 => 'a' \} \}; \#same hash as in prev example&#8203;: %hash =
> \( 1 => 'a' \)
> while\( my\( $key\, $value \) =  each %hash \) \{
>     say "$key $value";
> \}
> 
> \#test 5 FAIL not expected \(infinite loop\)
> while\( my\( $key\, $value \) =  each %\{ \{ 1 => 'a' \} \} \) \{
>     say "$key $value";
> \}
> 
> \#test 6 FAIL not Expected \(infinite loop\)
> while\( my\( $key\, $value \) =  each \{ 1 => 'a' \} \) \{
>     say "$key $value";
> \}
> 
> 

Since I have to go AFK now\, I'll leave analysis of these two cases to others\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From zefram@fysh.org

kes\-kes@&#8203;yandex\.ua wrote&#8203;:
> \#test 2 OK \(Why this work if "Type of argument to each on reference
> must be unblessed hashref or arrayref at \./test\.pl line 18\."
> while\( my\( $key\, $value \) =  each %$val \) \{

It works because this isn't "each on reference"\.  It's the old\-style
each on a hash\.

> \#test 5 FAIL not expected \(infinite loop\)
> while\( my\( $key\, $value \) =  each %\{ \{ 1 => 'a' \} \} \) \{

This creates a fresh hash each time round the loop\, via the \{\} constructor
expression\.  Iterating on a fresh hash first returns the first element\.
Because you throw the hash away before the next iteration\, you never
see that you've reached the end of that hash\.

> \#test 6 FAIL not Expected \(infinite loop\)
> while\( my\( $key\, $value \) =  each \{ 1 => 'a' \} \) \{

Same again\, this time using each on the reference\.

\-zefram

**Comment by p5pRT** [comment]
### From @demerphq

On 16 May 2015 at 09&#8203;:56\, KES \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  KES
> \# Please include the string&#8203;:  \[perl \#125205\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=125205 >
>
>
> This is a bug report for perl from kes\-kes@&#8203;yandex\.ru\,
> generated with the help of perlbug 1\.39 running under perl 5\.18\.2\.
>
> \#\!/usr/bin/perl
>
> use Modern&#8203;::Perl;
>
> my $val = \{ 1 => 'a' \};
>
> \#test 1 OK
> while\( my\( $key\, $value \) =  each %\{ $val \} \) \{
>     say "$key $value";
> \}
>
> \#test 2 OK \(Why this work if "Type of argument to each on reference must be unblessed hashref or arrayref at \./test\.pl line 18\."
> while\( my\( $key\, $value \) =  each %$val \) \{
>     say "$key $value";
> \}
>
> \#test 3 FAIL not Expected
> \#while\( my\( $key\, $value \) =  each \( 1 => 'a' \) \) \{
> \#    say "$key $value";
> \#\}
>
> \#test4 OK as Expected \(see perldoc each\)
> my %hash =  %\{ \{ 1 => 'a' \} \}; \#same hash as in prev example&#8203;: %hash =  \( 1 => 'a' \)
> while\( my\( $key\, $value \) =  each %hash \) \{
>     say "$key $value";
> \}
>
> \#test 5 FAIL not expected \(infinite loop\)
> while\( my\( $key\, $value \) =  each %\{ \{ 1 => 'a' \} \} \) \{
>     say "$key $value";
> \}
>
> \#test 6 FAIL not Expected \(infinite loop\)
> while\( my\( $key\, $value \) =  each \{ 1 => 'a' \} \) \{
>     say "$key $value";
> \}
>
>
> \_\_END\_\_
>
> May you tell me please where I can view status of this BUG report?

As far as I can tell\, and I think Zefram agrees\, this is not a bug\.

Yves

**Comment by p5pRT** [comment]
### From @kentfredric

>
> > May you tell me please where I can view status of this BUG report?
>
> As far as I can tell\, and I think Zefram agrees\, this is not a bug\.


At best\, one could potentially submit documentation for perlfunc/each to
clarify this case being not\-what\-you\-expected and why\.


\-\- 
Kent

\*KENTNL\* \- https://metacpan.org/author/KENTNL

**Comment by p5pRT** [comment]
### From @jkeenan

Since this is not a bug in Perl\, I am taking the ticket for the purpose of closing it within 7 days unless there is further discussion\.


\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @KES777

>This creates a fresh hash each time round the loop\, via the \{\} constructor
expression\.

I do not agree with you\.
http&#8203;://perldoc\.perl\.org/functions/keys\.html

for\( keys \{ 1 => 'a'\, 2 => 'b' \} \) \{
print "$\_\\n";
\}

As you say this must also creates a fresh hash\, but it does not\.
\( I do not belive that you create temp array for all keys in hash at once and than iterate over it\, because it may memory wasting on big hashes\) Reading doc carefully I see that you uses some iterator&#8203;: "Each hash or array has its own internal iterator\, accessed by each\, keys\, and values"\. So in case of 'keys'/'values' you \*\*create hash once\*\* and than \*\*iterate\*\* over hash keys/values using its internal iterator\.

>Because you throw the hash away before the next iteration\, you never
see that you've reached the end of that hash\.

Becase 'each' access internal iterator of hash in same manner as 'keys'/'values' do\. You can use this iterator\.

1\. What is the benifit of recalculating argument of 'each' for every 'while' iteration?

2\. Does this broken in compare to 5\.12? print shift split ' '\, '1 2';

3\. Can I force hash for \( 1 => 'a'\, 2 => 'b' \) without temporary variable? maybe like&#8203;:
while\( my\( $key\, $value \) =  each once \( 1 => 'a'\, 2 => 'b' \) \) \{
\.\.\.
\}

or force to not calculate 'each's  EXPR on 'while' iterations&#8203;:
while\( my\( $key\, $value \) =  each once \{ 1 => 'a'\, 2 => 'b' \} \) \{
\.\.\.
\}

**Comment by p5pRT** [comment]
### From zefram@fysh.org

KES wrote&#8203;:
>for\( keys \{ 1 => 'a'\, 2 => 'b' \} \) \{

Different case\.  A while loop evaluates its conditional expression each
time round the loop\.  Your while/each loop therefore evaluates the each \{\}
expression each time round the loop\, which creates a fresh hash each time\.
But a foreach loop evaluates its list expression only once per execution
of the loop as a whole\, to produce a list of elements to subsequently
iterate over\.  So this foreach/keys loop evaluates the keys \{\} expression
once\, creating one hash\.  The keys operator immediately iterates over the
entire hash to produce the list of keys\, and the hash is then discarded\.
\*Then\* the foreach iteration begins\, iterating over the temporarily\-saved
list of keys and executing the body once for each of those keys\.

>As you say this must also creates a fresh hash\, but it does not\.

It creates a fresh hash once per execution of the whole foreach loop\.
Totally different from the while loop\.

>\( I do not belive that you create temp array for all keys in hash
>at once and than iterate over it\, because it may memory wasting on
>big hashes\)

That's what it does\, wasteful or not\.  Except that technically we don't
call it an "array" because it's not reified as an array object; the
list of keys sits on an internal stack\.  It's an array in the C sense
but not in the Perl sense\.

>"Each hash or array has its own internal iterator\, accessed by each\,
>keys\, and values"\.

This refers to the well\-known side effect that the keys and values
operators interfere with each\, because they internally use the same
iterator\.  They reset the iterator\, use it to iterate over the entire
hash at once\, and leave the iterator reset again\.  This side effect is
sometimes used deliberately\, just to restart iteration that's using each\.

You can see this resetting effect by comparing these two loops&#8203;:

$ perl \-lwe '%a=\(a=>1\,b=>2\,c=>3\); while \($\_ = each %a\) \{ print $\_; \}'
c
a
b
$ perl \-lwe '%a=\(a=>1\,b=>2\,c=>3\); while \($\_ = each %a\) \{ print $\_; keys %a; \}' |head \-4
c
c
c
c

This means that there is practical utility in that \`wasteful' behaviour
of getting the entire key list up front&#8203;: it means that the hash iterator
can be safely used in other ways during the loop body\.  Not an issue in
your example where the hash is anonymous and not accessible outside the
loop control expression\, but I have on occasion fixed bugs by changing
a while/each loop into foreach/keys\.

>                   So in case of 'keys'/'values' you \*\*create hash once\*\*
>and than \*\*iterate\*\* over hash keys/values using its internal iterator\.

The keys operator does all the hash iteration in one go\.  With the choice
between foreach/keys and while/each\, you have a choice between doing all
the hash iteration up front and doing it as part of your loop structure\.
There's more than one way to do it\.

>1\. What is the benifit of recalculating argument of 'each' for every 'while' iteration?

It is fundamental to the while operator that it evaluates the conditional
expression anew each time round the loop\.  That implies recalculating
all the subexpressions\.  Consistency in these semantics is important to
make the language usable\.  Any subexpression that you don't want to be
re\-evaluated on each iteration needs to be lifted out of the loop\, as in&#8203;:

&nbsp;    my $h = \{ 1 => "a"\, 2 => "b" \};
&nbsp;    while\(my\($key\, $value\) = each %$h\) \{

>2\. Does this broken in compare to 5\.12? print shift split ' '\, '1 2';

shift is another of the operators to which the experimental implicit
dereferencing applies\.  That was introduced in Perl 5\.14\, which means
that the error message that you get \(and the time when you get it\)
has changed between 5\.12 and 5\.14\.  Neither Perl version is broken\,
but that expression is broken on any version\.

>3\. Can I force hash for \( 1 => 'a'\, 2 => 'b' \) without temporary variable? maybe like&#8203;:
>while\( my\( $key\, $value \) =  each once \( 1 => 'a'\, 2 => 'b' \) \) \{

That doesn't quite make sense\, because you don't have a hash constructor
expression there\.  But your other version&#8203;:

>or force to not calculate 'each's  EXPR on 'while' iterations&#8203;:
>while\( my\( $key\, $value \) =  each once \{ 1 => 'a'\, 2 => 'b' \} \) \{

is precisely possible\, using a module from CPAN&#8203;:

&nbsp;    use Memoize&#8203;::Once qw\(once\);
&nbsp;    while\( my\( $key\, $value \) =  each once \{ 1 => 'a'\, 2 => 'b' \} \) \{

See also the quite similar Memoize&#8203;::Lift\.  However\, these operators cause
the subexpression\, in this case the hash constructor\, to be executed
once \*per program run\*\, which may not be what you want\.  It's quite
different from the once per loop execution that you get from foreach/keys\.

\-zefram

**Comment by p5pRT** [comment]
### From @KES777

Ok\, thank you for your explanation\.
I agree that EXPR must be calculated for each iteration of while\, but\, I think\, it will be more sence if arguments of each will be calculated once

18\.05\.2015\, 19&#8203;:37\, "Zefram via RT" \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> KES wrote&#8203;:
>> for\( keys \{ 1 => 'a'\, 2 => 'b' \} \) \{
>
> Different case\.  A while loop evaluates its conditional expression each
> time round the loop\.  Your while/each loop therefore evaluates the each \{\}
> expression each time round the loop\, which creates a fresh hash each time\.
> But a foreach loop evaluates its list expression only once per execution
> of the loop as a whole\, to produce a list of elements to subsequently
> iterate over\.  So this foreach/keys loop evaluates the keys \{\} expression
> once\, creating one hash\.  The keys operator immediately iterates over the
> entire hash to produce the list of keys\, and the hash is then discarded\.
> \*Then\* the foreach iteration begins\, iterating over the temporarily\-saved
> list of keys and executing the body once for each of those keys\.
>> As you say this must also creates a fresh hash\, but it does not\.
>
> It creates a fresh hash once per execution of the whole foreach loop\.
> Totally different from the while loop\.
>> \( I do not belive that you create temp array for all keys in hash
>> at once and than iterate over it\, because it may memory wasting on
>> big hashes\)
>
> That's what it does\, wasteful or not\.  Except that technically we don't
> call it an "array" because it's not reified as an array object; the
> list of keys sits on an internal stack\.  It's an array in the C sense
> but not in the Perl sense\.
>> "Each hash or array has its own internal iterator\, accessed by each\,
>> keys\, and values"\.
>
> This refers to the well\-known side effect that the keys and values
> operators interfere with each\, because they internally use the same
> iterator\.  They reset the iterator\, use it to iterate over the entire
> hash at once\, and leave the iterator reset again\.  This side effect is
> sometimes used deliberately\, just to restart iteration that's using each\.
>
> You can see this resetting effect by comparing these two loops&#8203;:
>
> $ perl \-lwe '%a=\(a=>1\,b=>2\,c=>3\); while \($\_ = each %a\) \{ print $\_; \}'
> c
> a
> b
> $ perl \-lwe '%a=\(a=>1\,b=>2\,c=>3\); while \($\_ = each %a\) \{ print $\_; keys %a; \}' |head \-4
> c
> c
> c
> c
>
> This means that there is practical utility in that \`wasteful' behaviour
> of getting the entire key list up front&#8203;: it means that the hash iterator
> can be safely used in other ways during the loop body\.  Not an issue in
> your example where the hash is anonymous and not accessible outside the
> loop control expression\, but I have on occasion fixed bugs by changing
> a while/each loop into foreach/keys\.
>>                    So in case of 'keys'/'values' you \*\*create hash once\*\*
>> and than \*\*iterate\*\* over hash keys/values using its internal iterator\.
>
> The keys operator does all the hash iteration in one go\.  With the choice
> between foreach/keys and while/each\, you have a choice between doing all
> the hash iteration up front and doing it as part of your loop structure\.
> There's more than one way to do it\.
>> 1\. What is the benifit of recalculating argument of 'each' for every 'while' iteration?
>
> It is fundamental to the while operator that it evaluates the conditional
> expression anew each time round the loop\.  That implies recalculating
> all the subexpressions\.  Consistency in these semantics is important to
> make the language usable\.  Any subexpression that you don't want to be
> re\-evaluated on each iteration needs to be lifted out of the loop\, as in&#8203;:
>
>         my $h = \{ 1 => "a"\, 2 => "b" \};
>         while\(my\($key\, $value\) = each %$h\) \{
>> 2\. Does this broken in compare to 5\.12? print shift split ' '\, '1 2';
>
> shift is another of the operators to which the experimental implicit
> dereferencing applies\.  That was introduced in Perl 5\.14\, which means
> that the error message that you get \(and the time when you get it\)
> has changed between 5\.12 and 5\.14\.  Neither Perl version is broken\,
> but that expression is broken on any version\.
>> 3\. Can I force hash for \( 1 => 'a'\, 2 => 'b' \) without temporary variable? maybe like&#8203;:
>> while\( my\( $key\, $value \) =  each once \( 1 => 'a'\, 2 => 'b' \) \) \{
>
> That doesn't quite make sense\, because you don't have a hash constructor
> expression there\.  But your other version&#8203;:
>> or force to not calculate 'each's  EXPR on 'while' iterations&#8203;:
>> while\( my\( $key\, $value \) =  each once \{ 1 => 'a'\, 2 => 'b' \} \) \{
>
> is precisely possible\, using a module from CPAN&#8203;:
>
>         use Memoize&#8203;::Once qw\(once\);
>         while\( my\( $key\, $value \) =  each once \{ 1 => 'a'\, 2 => 'b' \} \) \{
>
> See also the quite similar Memoize&#8203;::Lift\.  However\, these operators cause
> the subexpression\, in this case the hash constructor\, to be executed
> once \*per program run\*\, which may not be what you want\.  It's quite
> different from the once per loop execution that you get from foreach/keys\.
>
> \-zefram

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @ap

\* KES \<kes\-kes@&#8203;yandex\.ua> \[2015\-05\-18 19&#8203;:20\]&#8203;:
> I agree that EXPR must be calculated for each iteration of while\, but\,
> I think\, it will be more sence if arguments of each will be calculated
> once

No\. That would require a combination of several special cases\, and that
would be worse than your surprise in this case\.

&nbsp;   “Any language that doesn't occasionally surprise the novice
&nbsp;    will pay for it by continually surprising the expert\.”   —Larry Wall

Now you know how it works\. So just store it in a variable and move on\.

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>
